src/main/java/com/google/sps/DataGraph.java

```diff
@@ -309,27 +315,49 @@ public String mutateGraph(Mutation mut) {
    * Modifies the list of tokens of this node to either add or remove tokens contained in tokenMut
    *
    * @param node the node whose token list should be modified
-   * @param tokenMut the mutation that should be applied to the token list
-   * @return the new GraphNode object, or null if it's an unrecognized mutation
+   * @param tokenMut the mutation that should be applied to the token list. The mutation should be
+   *     passed in builder form so that it can be deduplicated if necessary.
+   * @return the new GraphNode object, or null if it's an unrecognized mutation. The method may also
+   *     modify the mutation itself if it performs a duplicate action like adding an existing token.
    */
-  private GraphNode changeNodeToken(GraphNode node, TokenMutation tokenMut) {
+  private GraphNode changeNodeToken(GraphNode node, TokenMutation.Builder tokenMut) {
 
     // List of tokens to add/remove from the existing list
     List<String> tokenNames = tokenMut.getTokenNameList();
     // The existing list of tokens in the node
-    List<String> tokenList = new ArrayList<>();
-    tokenList.addAll(node.tokenList());
+    List<String> existingTokens = node.tokenList();
+    // The modified list of tokens of the node
+    Set<String> tokenList = new HashSet<>();
```

Comment:
could we change the variable name to tokenSet instead of tokenList since it's a set?

Reply:
I'll change it here, but I think the name in the proto is still tokenList :(.